### PR TITLE
Fixes #30. Support for shared library

### PR DESCRIPTION
**Pull request by machaval**
This pull request adds support for generating shared libraries by adding a new task.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -11,6 +11,9 @@ Gradle Tasks
 `./gradlew nativeImage`: create a native image using GraalVM's `native-image` tool with the configuration as specified
 by the `graal` Gradle extension. Outputs are produced to `${projectDir}/build/graal/`.
 
+`./gradlew sharedLibary`: * `sharedLibrary`: runs GraalVM's `native-image` for `shared library` command with configured
```
Comment (uschi2000) [inline README.md]: stray `*`

Comment (uschi2000) [inline README.md]: this sentence doesn't quite parse for me. how about:

like the `nativeImage` task, but configures the `native-image` tool to produce a shared library

File: src/main/java/com/palantir/gradle/graal/BaseGraalCompileTask.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    private final Property<Configuration> classpath = getProject().getObjects().property(Configuration.class);
+    private final RegularFileProperty jarFile = getProject().getObjects().fileProperty();
+    private final Property<Path> cacheDir = getProject().getObjects().property(Path.class);
+
+    public BaseGraalCompileTask() {
+        setGroup(GradleGraalPlugin.TASK_GROUP);
+        this.outputFile.set(getProject().getLayout().getBuildDirectory()
+                .dir("graal")
+                .map(d -> d.file(outputName.get())));
+
```
Comment (uschi2000) [inline src/main/java/com/palantir/gradle/graal/BaseGraalCompileTask.java]: remove 2x extra \n

File: src/main/java/com/palantir/gradle/graal/BaseGraalCompileTask.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    protected final String getExecutable() {
+        return cacheDir.get()
+                .resolve(Paths.get(graalVersion.get(), "graalvm-ce-" + graalVersion.get()))
+                .resolve(getArchitectureSpecifiedBinaryPath())
+                .toFile()
+                .getAbsolutePath();
+    }
+
+    protected final void loadArgs(List<String> args) throws IOException {
```
Comment (uschi2000) [inline src/main/java/com/palantir/gradle/graal/BaseGraalCompileTask.java]: would prefer to call this `configureArgs` and write a one-sentence javadoc

File: src/main/java/com/palantir/gradle/graal/SharedLibraryTask.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+import org.gradle.api.tasks.TaskAction;
+
+/**
+ * Runs GraalVM's native-image for shared library command with configured options and parameters.
+ */
+public class SharedLibraryTask extends BaseGraalCompileTask {
+
+    public SharedLibraryTask() {
+        setDescription(
+                "Runs GraalVM's native-image for shared library command with configured options and parameters."
```
Comment (uschi2000) [inline src/main/java/com/palantir/gradle/graal/SharedLibraryTask.java]: this sentence doesn't quite parse for me

**Review by uschi2000 — CHANGES_REQUESTED** [review]
Thanks for contributing this, @machaval ! As discussed in https://github.com/palantir/gradle-graal/pull/41/files , I think we need to deconflict the `mainClass` config option with the `sharedLibrary` task. One option would be to make the nativeImage task fail iff mainClass isn't configured, and make the sharedLibrary task fail iff mainClass is configured.

**Comment by uschi2000** [comment]
Fixes #39

File: src/main/java/com/palantir/gradle/graal/SharedLibraryTask.java (hunk 4 above)
Comment (machaval) [reply src/main/java/com/palantir/gradle/graal/SharedLibraryTask.java]: Would you help me with a more consistent sentence here. My english is not that good

File: README.md (hunk 1 above)
Comment (machaval) [reply README.md]: Can you help me with a better sentence. My english is not very good :(

File: src/main/java/com/palantir/gradle/graal/SharedLibraryTask.java (hunk 4 above)
Comment (uschi2000) [reply src/main/java/com/palantir/gradle/graal/SharedLibraryTask.java]: of course:
```
Runs GraalVM's native-image command configured to produce a shared library.
```
(same should go to the class javadoc)

File: README.md (hunk 1 above)
Comment (uschi2000) [reply README.md]: of course. you can just use the one I wrote above:

like the `nativeImage` task, but configures the `native-image` tool to produce a shared library

**Comment by machaval** [comment]
Just one comment on the mainClass. With the current approach it fails with the mainClass is not specified and the nativeImage task is executed, but it doesn't if the sharedLibrary is executed. This allows us to both generate sharedLibrary and nativeImage on a same gradle module.

Thanks for the fast feedback!!!!

**Review by uschi2000 — APPROVED** [review]
thanks!

**Comment by uschi2000** [comment]
This will be released as 0.3.0 as soon as the build finishes.

**Pull request merged**